src/main/java/com/aws/greengrass/clientdevices/auth/PermissionEvaluationUtils.java

```diff
@@ -132,21 +179,68 @@ private static Operation parseOperation(String operationStr) {
                 SERVICE_OPERATION_PATTERN.pattern()));
     }
 
-    private static Resource parseResource(String resourceStr) {
+    private Resource parseResource(String resourceStr) {
         if (Utils.isEmpty(resourceStr)) {
             throw new IllegalArgumentException("Resource can't be empty");
         }
 
         Matcher matcher = SERVICE_RESOURCE_PATTERN.matcher(resourceStr);
         if (matcher.matches()) {
-            return Resource.builder().service(matcher.group(1)).resourceType(matcher.group(2))
-                    .resourceName(matcher.group(3)).build();
+            return Resource.builder().service(matcher.group(1))
+                    .resourceType(matcher.group(2))
+                    .resourceName(matcher.group(3))
+                    .build();
         }
 
         throw new IllegalArgumentException(
                 String.format("Resource %s is not in the form of %s", resourceStr, SERVICE_RESOURCE_PATTERN.pattern()));
     }
 
+    /**
+     * utility method to replace policy variables in permission with device attribute.
+     *
+     * @param session    current device session
+     * @param permission permission to parse
+     * @return updated permission
+     */
+    Permission replaceResourcePolicyVariable(@NonNull Session session, @NonNull Permission permission) {
+
+        String resource = permission.getResource();
+
+        Matcher matcher = POLICY_VARIABLE_PATTERN.matcher(resource);
+
+        while (matcher.find()) {
+            String policyVariable = matcher.group(1);
+            String[] vars = policyVariable.split("\\.");
+            String attributeNamespace = vars[1];
+            String attributeName = vars[2];
+
+            // this supports the ThingName attribute only
+            if (THING_NAME_VARIABLE.equals(policyVariable)) {
+                String policyVariableValue =
+                        Coerce.toString(session.getSessionAttribute(attributeNamespace, attributeName));
+
+                if (policyVariableValue == null) {
+                    logger.atWarn().kv("attributeName", attributeName)
+                            .log("No attribute found for current session");
+                } else {
+                    // for ThingName support only, we can use .replaceAll()
+                    // to support additional policy variables in the future
+                    resource = matcher.replaceFirst(policyVariableValue);
+                    matcher = POLICY_VARIABLE_PATTERN.matcher(resource);
+                    permission = Permission.builder()
+                            .principal(permission.getPrincipal())
+                            .operation(permission.getOperation())
+                            .resource(resource).build();
+                }
+            } else {
+                logger.atWarn().kv("policyVariable", policyVariable)
+                        .log("Policy variable detected but could not be parsed");
```

Comment:
policy variable `unsupported`,  and can also guide user w/ next steps (e.g. fix config)